### PR TITLE
Redefine AWS_PRECONDITION and add an explanation on existing preconditions

### PR DESCRIPTION
**Pull request by angelhof**
Redefine `AWS_PRECONDITION` using `__CPROVER_precondition` so that it includes an explanation message.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/array_list.inl [hunk 1]
```diff
@@ -91,7 +91,7 @@ bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
 
 AWS_STATIC_IMPL
 void aws_array_list_clean_up(struct aws_array_list *AWS_RESTRICT list) {
-    AWS_PRECONDITION(aws_array_list_is_valid(list));
+    AWS_PRECONDITION(aws_array_list_is_valid(list), "Input array_list [list] should be valid.");
```
Comment (danielsn) [inline include/aws/common/array_list.inl]: I wonder if "must" is better than "should" here.

File: include/aws/common/common.h [hunk 2]
```diff
@@ -117,12 +117,12 @@
  * Violations of the function contracts are undefined behaviour.
  */
 #ifdef CBMC
-#define AWS_PRECONDITION(cond) __CPROVER_assume(cond)
+#define AWS_PRECONDITION(cond, explanation) __CPROVER_precondition((cond), (explanation))
```
Comment (danielsn) [inline include/aws/common/common.h]: Might make sense to append "Precondition: " to the string?

File: source/hash_table.c [hunk 3]
```diff
@@ -53,10 +53,10 @@ static uint64_t s_hash_for(struct hash_table_state *state, const void *key) {
 }
 
 static size_t s_index_for(struct hash_table_state *map, struct hash_table_entry *entry) {
-    AWS_PRECONDITION(hash_table_state_is_valid(map));
+    AWS_PRECONDITION(hash_table_state_is_valid(map), "Input hash_table_state [map] should be valid.");
     size_t index = entry - map->slots;
     AWS_POSTCONDITION(index < map->size);
-    AWS_PRECONDITION(hash_table_state_is_valid(map));
+    AWS_POSTCONDITION(hash_table_state_is_valid(map));
```
Comment (danielsn) [inline source/hash_table.c:59]: Nice catch.

File: source/hash_table.c [hunk 4]
```diff
@@ -195,7 +195,9 @@ int aws_hash_table_init(
     aws_hash_callback_eq_fn *equals_fn,
     aws_hash_callback_destroy_fn *destroy_key_fn,
     aws_hash_callback_destroy_fn *destroy_value_fn) {
-    AWS_PRECONDITION(map && alloc && hash_fn && equals_fn);
+    AWS_PRECONDITION(
+        map && alloc && hash_fn && equals_fn,
+        "Input pointers [map], [alloc], [hash_fn], and [equals_fn] shouldn't be NULL.");
```
Comment (danielsn) [inline source/hash_table.c]: Would it be cleaner to make this 4 preconditions?

File: source/hash_table.c [hunk 5]
```diff
@@ -235,14 +237,16 @@ void aws_hash_table_clean_up(struct aws_hash_table *map) {
...
     struct aws_hash_table tmp = *a;
     *a = *b;
     *b = tmp;
 }
 
 void aws_hash_table_move(struct aws_hash_table *AWS_RESTRICT to, struct aws_hash_table *AWS_RESTRICT from) {
-    AWS_PRECONDITION(to != NULL && from != NULL && to != from && aws_hash_table_is_valid(from));
+    AWS_PRECONDITION(
+        to && from && to != from && aws_hash_table_is_valid(from),
+        "Input hash tables [to] and [from] should be different and [from] should also be valid.");
```
Comment (danielsn) [inline source/hash_table.c]: Same as above.  Might be cleaner to sepearte this into several conditions.

**Review by danielsn — APPROVED** [review]

File: include/aws/common/array_list.inl (hunk 1 above)
Comment (angelhof) [reply include/aws/common/array_list.inl]: I think I agree. "Should" sounds less strict than "must".

File: include/aws/common/common.h (hunk 2 above)
Comment (angelhof) [reply include/aws/common/common.h]: I agree :)

File: source/hash_table.c (hunk 5 above)
Comment (angelhof) [reply source/hash_table.c]: I agree for both the above.

**Pull request closed without merging**